File: backend-nestjs/src/seed.ts
import { NestFactory } from '@nestjs/core';
import { AppModule } from './app.module';
import { VideoService } from './video/video.service';
import { AnnotationService } from './annotation/annotation.service';
import { promises as fs } from 'fs';
import { join } from 'path';

async function seed() {
  const app = await NestFactory.createApplicationContext(AppModule);

  const videoService = app.get(VideoService);
  const annotationService = app.get(AnnotationService);

  try {
    // Create videos directory if it doesn't exist
    const videosDir = join(process.cwd(), 'videos');
    await fs.mkdir(videosDir, { recursive: true });

    // Sample video data
    const sampleVideos = [
      {
        title: 'Introduction to React Hooks',
        description:
          'Learn the fundamentals of React Hooks including useState, useEffect, and custom hooks.',
        filename: 'react-hooks-intro.mp4',
        originalName: 'react-hooks-intro.mp4',
        mimeType: 'video/mp4',
        size: 15728640, // 15MB
        duration: 1200, // 20 minutes
        views: 0,
        isActive: true,
      },
      {
        title: 'Advanced TypeScript Patterns',
        description:
          'Explore advanced TypeScript patterns and best practices for large-scale applications.',
        filename: 'typescript-advanced.mp4',
        originalName: 'typescript-advanced.mp4',
        mimeType: 'video/mp4',
        size: 25165824, // 24MB
        duration: 1800, // 30 minutes
        views: 0,
        isActive: true,
      },
      {
        title: 'GraphQL with NestJS',
        description: 'Build scalable APIs with GraphQL and NestJS framework.',
        filename: 'graphql-nestjs.mp4',
        originalName: 'graphql-nestjs.mp4',
        mimeType: 'video/mp4',
        size: 31457280, // 30MB
        duration: 2400, // 40 minutes
        views: 0,
        isActive: true,
      },
    ];

    console.log('🌱 Seeding videos...');
    const createdVideos = [];

    for (const videoData of sampleVideos) {
      const video = await videoService.create(videoData);
      createdVideos.push(video);
      console.log(`✅ Created video: ${video.title}`);
    }

    // Sample annotations for the first video
    const sampleAnnotations = [
      {
        title: 'Introduction',
        description: 'Welcome and overview of the course',
        startTime: 0,
        endTime: 60,
        type: 'chapter',
        color: '#3B82F6',
        videoId: createdVideos[0].id,
      },
      {
        title: 'useState Hook',
        description: 'Understanding the useState hook for state management',
        startTime: 60,
        endTime: 300,
        type: 'chapter',
        color: '#10B981',
        videoId: createdVideos[0].id,
      },
      {
        title: 'useEffect Hook',
        description: 'Side effects and lifecycle management with useEffect',
        startTime: 300,
        endTime: 600,
        type: 'chapter',
        color: '#F59E0B',
        videoId: createdVideos[0].id,
      },
      {
        title: 'Custom Hooks',
        description: 'Creating reusable custom hooks',
        startTime: 600,
        endTime: 900,
        type: 'chapter',
        color: '#EF4444',
        videoId: createdVideos[0].id,
      },
      {
        title: 'Best Practices',
        description: 'React Hooks best practices and common pitfalls',
        startTime: 900,
        endTime: 1200,
        type: 'chapter',
        color: '#8B5CF6',
        videoId: createdVideos[0].id,
      },
    ];

    console.log('🌱 Seeding annotations...');
    for (const annotationData of sampleAnnotations) {
      const annotation = await annotationService.create(annotationData);
      console.log(`✅ Created annotation: ${annotation.title}`);
    }

    // Create sample video files
    console.log('📁 Creating sample video files...');
    const sampleVideoUrl =
      'https://www.learningcontainer.com/wp-content/uploads/2020/05/sample-mp4-file.mp4';

    for (const video of createdVideos) {
      const videoPath = join(videosDir, video.filename);
      try {
        await fs.access(videoPath);
        console.log(`📹 Video file already exists: ${video.filename}`);
      } catch {
        try {
          // Download sample video file
          const response = await fetch(sampleVideoUrl);
          if (response.ok) {
            const buffer = await response.arrayBuffer();
            await fs.writeFile(videoPath, Buffer.from(buffer));
            console.log(`📹 Created sample video file: ${video.filename}`);
          } else {
            // Fallback: create a minimal valid MP4 file
            const minimalMp4 = Buffer.from([
              0x00, 0x00, 0x00, 0x20, 0x66, 0x74, 0x79, 0x70, 0x69, 0x73, 0x6f,
              0x6d, 0x00, 0x00, 0x02, 0x00, 0x69, 0x73, 0x6f, 0x6d, 0x69, 0x73,
              0x6f, 0x32, 0x61, 0x76, 0x63, 0x31, 0x6d, 0x70, 0x34, 0x31,
            ]);
            await fs.writeFile(videoPath, minimalMp4);
            console.log(`📹 Created minimal MP4 file: ${video.filename}`);
          }
        } catch (error) {
          console.warn(
            `⚠️  Could not create video file ${video.filename}:`,
            error.message
          );
        }
      }
    }

    console.log('🎉 Seeding completed successfully!');
  } catch (error) {
    console.error('❌ Seeding failed:', error);
  } finally {
    await app.close();
  }
}

seed();

File: backend-fastapi/seed.py
"""
Database seeding script for FastAPI backend.
"""

import asyncio
import os
from pathlib import Path
import aiofiles
from sqlalchemy.ext.asyncio import AsyncSession
from app.core.database import AsyncSessionLocal, engine, Base
from app.models import Video, Annotation, VideoChunk
from app.services.video_service import VideoService
from app.services.annotation_service import AnnotationService
from app.services.video_processing import VideoProcessingService
from app.schemas.video import VideoCreate
from app.schemas.annotation import AnnotationCreate


async def seed_database():
    """Seed the database with sample data."""
    print("🌱 Starting FastAPI database seeding...")
    
    # Create database tables
    async with engine.begin() as conn:
        await conn.run_sync(Base.metadata.create_all)
    
    # Create videos directory
    videos_dir = Path("videos")
    videos_dir.mkdir(exist_ok=True)
    
    async with AsyncSessionLocal() as db:
        video_service = VideoService(db)
        annotation_service = AnnotationService(db)
        
        # Sample videos data - Only demo video for YouTube-like player
        sample_videos = [
            {
                "title": "Demo Video - YouTube-like Player",
                "description": "A demonstration video showcasing the YouTube-like video player with frame preview and chunking capabilities. This 30-minute video demonstrates how large video files are automatically chunked for efficient streaming.",
                "filename": "demo_video.mp4",
                "originalName": "demo_video.mp4",
                "mimeType": "video/mp4",
                "size": 52428800,  # 50MB
                "duration": 1800,  # 30 minutes for chunking demo
            },
        ]
        
        # Check if videos already exist
        existing_videos = await video_service.get_all()
        if existing_videos:
            print("⚠️ Videos already exist in database. Skipping seeding to prevent duplicates.")
            return
        
        # Create videos
        print("🌱 Seeding videos...")
        created_videos = []
        
        for video_data in sample_videos:
            video_create = VideoCreate(**video_data)
            video = await video_service.create(video_create)
            created_videos.append(video)
            print(f"✅ Created video: {video.title}")
        
        # Sample annotations for the 30-minute demo video
        sample_annotations = [
            {
                "title": "Video Player Introduction",
                "description": "Introduction to the YouTube-like video player features",
                "startTime": 0,
                "endTime": 180,  # 3 minutes
                "type": "chapter",
                "color": "#3B82F6",
                "videoId": created_videos[0].id,
            },
            {
                "title": "Frame Preview Demo",
                "description": "Demonstration of frame preview on timeline hover",
                "startTime": 180,
                "endTime": 360,  # 3-6 minutes
                "type": "chapter",
                "color": "#10B981",
                "videoId": created_videos[0].id,
            },
            {
                "title": "Video Chunking System",
                "description": "How video chunking works for large files",
                "startTime": 360,
                "endTime": 720,  # 6-12 minutes
                "type": "chapter",
                "color": "#F59E0B",
                "videoId": created_videos[0].id,
            },
            {
                "title": "Keyboard Shortcuts",
                "description": "Advanced keyboard controls for video navigation",
                "startTime": 720,
                "endTime": 1080,  # 12-18 minutes
                "type": "chapter",
                "color": "#EF4444",
                "videoId": created_videos[0].id,
            },
            {
                "title": "Annotation System",
                "description": "Adding and managing video annotations",
                "startTime": 1080,
                "endTime": 1440,  # 18-24 minutes
                "type": "chapter",
                "color": "#8B5CF6",
                "videoId": created_videos[0].id,
            },
            {
                "title": "Performance Features",
                "description": "Optimizations for large video files",
                "startTime": 1440,
                "endTime": 1800,  # 24-30 minutes
                "type": "chapter",
                "color": "#06B6D4",
                "videoId": created_videos[0].id,
            },
        ]
        
        # Create annotations
        print("🌱 Seeding annotations...")
        for annotation_data in sample_annotations:
            annotation_create = AnnotationCreate(**annotation_data)
            annotation = await annotation_service.create(annotation_create)
            print(f"✅ Created annotation: {annotation.title}")
        
        # Create sample video files
        print("📁 Creating sample video files...")
        sample_video_url = "https://www.learningcontainer.com/wp-content/uploads/2020/05/sample-mp4-file.mp4"
        
        for video in created_videos:
            video_path = videos_dir / video.filename
            if not video_path.exists():
                try:
                    import aiohttp
                    async with aiohttp.ClientSession() as session:
                        async with session.get(sample_video_url) as response:
                            if response.status == 200:
                                content = await response.read()
                                async with aiofiles.open(video_path, 'wb') as f:
                                    await f.write(content)
                                print(f"📹 Created sample video file: {video.filename}")
                            else:
                                print(f"⚠️ Could not download sample video for {video.filename}")
                except Exception as e:
                    print(f"⚠️ Error creating video file {video.filename}: {e}")
            else:
                print(f"📹 Video file already exists: {video.filename}")
        
        # Process video chunks for the demo video
        print("🔧 Processing video chunks...")
        video_processing_service = VideoProcessingService(chunk_duration=120)  # 2 minutes
        
        for video in created_videos:
            video_path = videos_dir / video.filename
            if video_path.exists() and video.title == "Demo Video - YouTube-like Player":
                try:
                    chunks = await video_processing_service.chunk_video(video.id, video_path, db)
                    print(f"✅ Created {len(chunks)} chunks for {video.title}")
                except Exception as e:
                    print(f"⚠️ Error chunking video {video.title}: {e}")
        
        print("🎉 FastAPI seeding completed successfully!")


if __name__ == "__main__":
    asyncio.run(seed_database())
